### PR TITLE
definition of splitted contexts

### PR DESCRIPTION
**Pull request by koseoyoung**
works on #1689 
First, define the new contexts which are classified by their groups

### CODE REVIEW

File: action/protocol/context.go [hunk 1]
```diff
@@ -98,7 +98,115 @@ func GetValidateActionsCtx(ctx context.Context) (ValidateActionsCtx, bool) {
...
+}
+
+// WithBlockchainCtx add BlockchainCtx into context.
+func WithBlockchainCtx(ctx context.Context, bc BlockchainCtx) context.Context {
+	return context.WithValue(ctx, blockchainCtxKey{}, bc)
+}
+
+// GetBlockchainCtx gets BlockchainCtx
+func GetBlockchainCtx(ctx context.Context) (BlockchainCtx, bool) {
+	bc, ok := ctx.Value(blockchainCtxKey{}).(BlockchainCtx)
```
Comment (CoderZhi) [inline action/protocol/context.go]: `return ctx.Value(blockchainCtxKey{}).(BlockchainCtx)`

File: action/protocol/context.go [hunk 2]
```diff
@@ -98,7 +98,115 @@ func GetValidateActionsCtx(ctx context.Context) (ValidateActionsCtx, bool) {
...
+}
+
+// WithBlockCtx add BlockCtx into context.
+func WithBlockCtx(ctx context.Context, blk BlockCtx) context.Context {
+	return context.WithValue(ctx, blockCtxKey{}, blk)
+}
+
+// GetBlockCtx gets BlockCtx
+func GetBlockCtx(ctx context.Context) (BlockCtx, bool) {
+	blk, ok := ctx.Value(blockCtxKey{}).(BlockCtx)
```
Comment (CoderZhi) [inline action/protocol/context.go]: dito

File: action/protocol/context_test.go [hunk 3]
```diff
@@ -82,5 +82,102 @@ func TestMustGetValidateActionsCtx(t *testing.T) {
 	ret := MustGetValidateActionsCtx(ctx)
 	require.Equal(uint64(1111), ret.BlockHeight)
 	// Case II: Panic
-	require.Panics(func() { MustGetValidateActionsCtx(context.Background()) }, "Miss run actions context")
+	require.Panics(func() { MustGetValidateActionsCtx(context.Background()) }, "Miss validate action context")
+}
+
+func TestWithBlockchainCtx(t *testing.T) {
+	require := require.New(t)
+	bcCtx := BlockchainCtx{config.Default.Genesis, false, nil}
```
Comment (CoderZhi) [inline action/protocol/context_test.go]:
```
    bcCtx := BlockchainCtx{
        Genesis: config.Default.Genesis,
        History: false,
        Registry: nil,
    }
```

Same for the rest

File: action/protocol/context.go [hunk 4]
```diff
@@ -98,7 +98,115 @@ func GetValidateActionsCtx(ctx context.Context) (ValidateActionsCtx, bool) {
...
 	if !ok {
-		log.S().Panic("Miss run actions context")
+		log.S().Panic("Miss validate actions context")
 	}
 	return va
 }
+
+// TODO: replace RunActionsCtx and ValidateActionsCtx with below classified independent contexts
+
+type blockchainCtxKey struct{}
```
Comment (CoderZhi) [inline action/protocol/context.go]: let's replace "Ctx" with "Context"

File: action/protocol/context_test.go (hunk 3 above)
Comment (koseoyoung) [reply action/protocol/context_test.go]: fixed in latest commit

File: action/protocol/context.go (hunk 1 above)
Comment (koseoyoung) [reply action/protocol/context.go]: interface type assertion needs to be assigned to return 2 values

File: action/protocol/context.go (hunk 4 above)
Comment (koseoyoung) [reply action/protocol/context.go]: fixed in latest commit

**Pull request merged**